File: src/theme.ts
import { createColors } from "picocolors";
import type { WorktreeView } from "./types";

export interface TerminalTheme {
  header(value: string): string;
  current(value: string): string;
  branch(value: string): string;
  detached(value: string): string;
  commit(value: string): string;
  path(value: string, selected: boolean): string;
  status(value: string, worktree: WorktreeView): string;
  selection(value: string): string;
  hint(value: string): string;
  key(value: string): string;
}

export function createTerminalTheme(color: boolean): TerminalTheme {
  const colors = createColors(color);
  const boldWhite = (value: string) => colors.bold(colors.white(value));
  const boldCyan = (value: string) => colors.bold(colors.cyan(value));

  return {
    header: boldWhite,
    current: (value) => colors.bold(colors.green(value)),
    branch: colors.cyan,
    detached: colors.yellow,
    commit: colors.dim,
    path: (value, selected) => selected ? colors.bold(value) : value,
    status: (value, worktree) => {
      if (worktree.availability === "prunable" || worktree.availability === "unavailable") {
        return colors.red(value);
      }
      if (worktree.availability === "bare") return colors.yellow(value);
      if (!worktree.status) return colors.red(value);

      const parts: string[] = [];
      if (worktree.status.conflicted) parts.push(colors.red(`${worktree.status.conflicted} conflicted`));
      if (worktree.status.staged) parts.push(colors.yellow(`${worktree.status.staged} staged`));
      if (worktree.status.modified) parts.push(colors.blueBright(`${worktree.status.modified} modified`));
      if (worktree.status.untracked) parts.push(colors.red(`${worktree.status.untracked} untracked`));
      if (worktree.locked !== undefined) {
        parts.push(colors.yellow(worktree.locked ? `locked: ${worktree.locked}` : "locked"));
      }
      return parts.length ? parts.join(", ") : colors.green(value);
    },
    selection: boldCyan,
    hint: colors.dim,
    key: colors.bold,
  };
}

File: src/theme.rs
use crate::{
    status,
    types::{Availability, Worktree},
};
pub struct Theme {
    pub color: bool,
}
impl Theme {
    fn wrap(&self, code: &str, s: &str) -> String {
        if self.color {
            format!("\x1b[{code}m{s}\x1b[0m")
        } else {
            s.into()
        }
    }
    pub fn header(&self, s: &str) -> String {
        self.wrap("1;37", s)
    }
    pub fn current(&self, s: &str) -> String {
        self.wrap("1;32", s)
    }
    pub fn branch(&self, s: &str) -> String {
        self.wrap("36", s)
    }
    pub fn occupied_branch(&self, s: &str) -> String {
        self.wrap("33", s)
    }
    pub fn detached(&self, s: &str) -> String {
        self.wrap("33", s)
    }
    pub fn commit(&self, s: &str) -> String {
        self.wrap("2", s)
    }
    pub fn selected(&self, s: &str) -> String {
        self.wrap("1;36", s)
    }
    pub fn hint(&self, s: &str) -> String {
        self.wrap("2", s)
    }
    pub fn key(&self, s: &str) -> String {
        self.wrap("1", s)
    }
    pub fn wrap_selected_path(&self, s: &str) -> String {
        self.wrap("1", s)
    }
    pub fn status(&self, w: &Worktree) -> String {
        if !self.color {
            return status::format(w);
        }
        match w.availability {
            Availability::Unavailable | Availability::Prunable => {
                self.wrap("31", &status::format(w))
            }
            Availability::Bare => self.wrap("33", "bare"),
            Availability::Available => {
                let Some(c) = &w.status else {
                    return self.wrap("31", "unavailable");
                };
                let mut p = Vec::new();
                if c.conflicted > 0 {
                    p.push(self.wrap("31", &format!("{} conflicted", c.conflicted)))
                }
                if c.staged > 0 {
                    p.push(self.wrap("33", &format!("{} staged", c.staged)))
                }
                if c.modified > 0 {
                    p.push(self.wrap("94", &format!("{} modified", c.modified)))
                }
                if c.untracked > 0 {
                    p.push(self.wrap("31", &format!("{} untracked", c.untracked)))
                }
                if let Some(r) = &w.locked {
                    let label = if r.is_empty() {
                        "locked".into()
                    } else {
                        format!("locked: {r}")
                    };
                    p.push(self.wrap("33", &label))
                }
                if p.is_empty() {
                    self.wrap("32", "clean")
                } else {
                    p.join(", ")
                }
            }
        }
    }
}
